cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -268,35 +267,79 @@ protected void triggerShutdown() {
    * @param messageIds a map with each key {@link TopicIdMetaKey} representing a topic and messageId's
    *                   of the last persisted metric of the topic
    */
-  private void persistMetricsMessageIds(Deque<MetricValues> metricValues, Map<TopicIdMetaKey, byte[]> messageIds) {
+  private void persistMetricsMessageIds(Deque<MetricValues> metricValues,
+                                        Map<TopicIdMetaKey, MetricsProcessorStats> messageIds) {
     try {
+      // we update the processing stats in meta table before we persist metrics,
+      // in case of huge-delay in writing to metrics table (maybe due to region server issues),
+      // this will help surface the last processed timestamp information to the users.
+      updateProcessingStats(messageIds);
       if (!metricValues.isEmpty()) {
-        persistMetrics(metricValues);
+        persistMetrics(metricValues, messageIds);
       }
       persistMessageIds(messageIds);
+      // reset processing stats, so the latest processing stats will get updated in the next iteration.
+      resetProcessingStats(messageIds);
```

Comment:
should this be done inside updateProcessingStats? otherwise, if something fails in between, the stats will not be reset